src/ng/directive/as.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * @ngdoc directive
+ * @name as
+ * @restrict A
+ *
+ * @description
+ * Publishes a component controller or a DOM element into the current scope.
+ *
+ * @element ANY
+ * @param {expression} as {@link guide/expression Expression} to assign the controller.
+ *
+ * This directive allows to access to the controller of components in your
+ * current template or from current controller.
+ *
+ * @example
+ * ### Access to component controller
+ * This example demonstrates basic publishing of the component controller into the scope.
+ * <example name="asDirectiveComponentExample" module="asComponentExample">
+ *   <file name="index.html">
+ *     <script>
+ *       angular.module('asComponentExample', [])
+ *        .component('toggle', {
+ *           controller: function() {
+ *             var opened = false;
+ *             this.isOpen = function() { return opened; };
+ *             this.toggle = function() { opened = !opened; };
+ *           }
+ *         });
+ *     </script>
+ *     <toggle as="myToggle"></toggle>
+ *     <button ng-click="myToggle.toggle()">Toggle</button>
+ *     <div ng-show="myToggle.isOpen()">You are using a children component to show it.</div>
+ *   </file>
+ *   <file name="protractor.js" type="protractor">
+ *      it('should publish the toggle into the scope', function() {
+ *        var toggle = element(by.buttonText('Toggle'));
+ *        expect(toggle.evaluate('myToggle.isOpen()')).toEqual(false);
+ *        toggle.click();
+ *        expect(toggle.evaluate('myToggle.isOpen()')).toEqual(true);
+ *      });
+ *   </file>
+ * </example>
+ *
+ * This directive also allows to bind elements to the scope.
+ *
+ * ```js
+ *   var myMod = angular.module(...);
+ *   myMod.component('myPlayer', {
+ *     template: '' +
+ *       '<audio as="$ctrl.audio" src="foo.ogg"></audio>' +
+ *       '<button ng-click="$ctrl.play()">Play</button>' +
+ *       '',
+ *     controller: function() {
+ *       this.play = function() {
+ *         this.audio.play();
+ *       };
+ *     }
+ *   });
+ * ```
+ *
+ * Note that angular does not allows to access DOM elements directly
+ * from templates, you need a controller to manipulate them.
+ *
+ */
+var componentAsDirective = ['$parse', function($parse) {
+  return {
+    restrict: 'A',
+    compile: function(tElement, tAttrs) {
+      // gets the theoretical controller name, converts <some-thing> into "someThing")
+      var controllerName = tElement[0].tagName.toLowerCase()
+        .replace(/[^\w]\w/g, function(a) { return a.slice(1).toUpperCase(); });
+
+      // get the setter for the as attribute
+      var setter = $parse(tAttrs.as).assign;
+
+      return function(scope, iElement, iAttrs) {
+        // gets the controller of the current element (see jqLiteController for details)
+        var controller = iElement.data('$' + controllerName + 'Controller');
+
+        if (controller) {
+          setter(scope, controller);
+        } else {
+          setter(scope, iElement[0]);
```

Comment:
> NG2 does something like this (it is not scope, but binds DOMElements to Controllers)

Angular 2 has better isolation, so it's not as much of a concern. Here we are attaching it on scope and there is the scope hierarchy, meaning that malicious code could reach the DOM element from virtually anyware on the page.
So, we don't have a way to have "template-local" variables as in ng2.

--
> Security issues are controlled by parser

Only when expensive checks are on, but we can't controller in what kind of expressions the malicious code would end up to. So, the parser doesn't have our back here.

--
> the nice alternative is to make nice small directives to wrap native elements like audio/video/

Sounds about right :smiley: